### ISSUE TITLE
Respect .yarnrc.yml registry config

### ISSUE THREAD

**Issue by rdicroce** [renovatebot/renovate#7351]
**What would you like Renovate to be able to do?**

Respect the registry configuration (URLs and auth settings) in .yarnrc.yml (Yarn V2). This is related to #5230, and is the same feature as #5821 except for Yarn V2 instead of V1.

**Did you already have any implementation ideas?**

It looks like getDependency in datasource/npm/get.ts would be the appropriate place to add the necessary code. The config could be resolved using Yarn's own APIs to ensure compatibility (see https://github.com/yarnpkg/berry/issues/264).

But I feel like the overall architecture needs some work. In particular, I think the datasource API needs some enhancement. Ideally it would have a method that would be called once per repository to allow the datasource to discover relevant config files. This method would return a datasource-specific object, and that object would be passed to getReleases() as a new parameter. That way, the npm datasource could look around the repo for .npmrc/.yarnrc/.yarnrc.yml/etc and then parse them appropriately to determine registry configuration. This would have the added benefit that the registry configuration would be applied for any dependency regardless of which manager discovered it, so there are no surprises if e.g. the regex manager discovers a dependency that uses the npm datasource.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

#5821 mentions duplicating the registry settings in .npmrc, since Renovate does check that file. But that's obviously not ideal.

**Is this a feature you'd be interested in implementing yourself?**

I'm willing to take a stab at it, though I'd want feedback about the implementation ideas first.

**Comment by rarkins** [comment]
It sounds like this should be using the `registryUrls` field which is passed to datasource, and best extracted as part of manager extract and not datasource get.

**Comment by rdicroce** [comment]
I thought about that but I see at least 2 problems with that approach:
1) URL is applied but auth config is not.
2) Nothing is applied if managers other than npm (e.g. regex) generate a dependency for the npm datasource.

**Comment by viceice** [comment]
The preferred way for renovate auth to data are hostRules.

So reading auth from npmrc, yarn ... is only a fallback.

**Comment by viceice** [comment]
Currently if we find auth in npmrc, we add them to hostrules transparently.

**Comment by rarkins** [comment]
Yes, `packageRules` to control where lookups go (e.g. if they apply only to certain scopes of package) and `hostRules` for auth. I assume you're not actually committing auth secrets to your repository yaml files?

**Comment by rdicroce** [comment]
No, in fact I don't need auth secrets at all in my repos, but other people might.

So the expectation is that registries and auth params should be set in the Renovate config? I'm okay with that, it's just that the current docs for registryUrls imply otherwise:

`This is only necessary in case you need to manually configure a registry URL to use for datasource lookups. Applies to PyPI (pip) only for now. Supports only one URL for now but is defined as a list for forward compatibility.`

So:
- Does this setting work for managers other than pip?
- Does it actually support more than 1 URL?

If yes to the above, then I'd suggest the description should be clarified to explain what situations make it "necessary" to configure the URLs. Or maybe there should be a dedicated page explaining how to configure Renovate for use with private datasources. There is already such a page for private NPM modules, but it's very NPM specific and makes no references to registryUrls or hostRules.

I can open a PR to add/update the docs if you'd like.

**Comment by rarkins** [comment]
For end users, `hostRules` are used for configuring host-specific parameters such as authentication. `packageRules` are used for configuring package-specific parameters which could also include prefix -> registryUrls mappings.

> So the expectation is that registries and auth params should be set in the Renovate config?

Auth is ideally done via config (because committing raw secrets to a repo is almost always a bad idea) but we have a few scenarios where we do read it from file (including `.npmrc`). Typically auth from files is a huge headache and we recommend `hostRules`. 

End users should not usually need to manually configure `registryUrls` manually because they can usually be detected by the manager extract function. I've pushed a fix to that `registryUrls` documentation just now.

Please note though that when I was describing `registryUrls` earlier it was in the context of *writing code*, not *being an end user configuring Renovate*. Ideally the npm/yarn extract function should detect the yarn config and include all necessary `registryUrls` config in its extracted results.

Any non-manager-specific page for how to configure private datasources would be great. We probably don't need an npm-specific one and could roll that up into the generic one.

**Comment by rdicroce** [comment]
Ah, okay, I see what you're saying now. I was getting thrown off because .npmrc handling is currently done by the datasource rather than the manager.

If I'm understanding the code correctly, the best solution would be to set registryUrls appropriately in each PackageDependency. Setting registryUrls in PackageFile wouldn't be correct since .yarnrc.yml allows different registries to be set for different scopes.

We don't need to do anything for lockfile maintenance since that is accomplished by invoking Yarn.

But there's still the problem of another manager generating a dependency for the npm datasource. In particular, I have a case where I'm using the regex manager with the npm datasource. Renovate has to know which registry to use for that somehow. Any thoughts?

**Comment by viceice** [comment]
It should be possible to use a package rule with manager, file and dependency selector to assign additional settings like registry url

**Comment by rdicroce** [comment]
Right, obviously that can be done by manually adding packageRules in the Renovate config. But that means users have to duplicate the registry settings, which is less than ideal and IMO is likely to surprise users. They have a .yarnrc.yml (or a .npmrc or whatever) and they have a file (in the same directory or subdirectory) that contains a npm depdendency. As a user, I would expect my Yarn/NPM/etc registry settings to apply to that dependency. And for .npmrc, that is the current behavior, because the registry URL is determined by the datasource rather than the manager.

Unless you mean that the manager should be able to return a list of packageRules which are then automatically added to the packageRules supplied by the user? It doesn't look like managers can currently do that.

**Comment by viceice** [comment]
The normal renovate flow should be that the manager extract the dependency, add optional registry url if configured. Eg read npmrc.

And the datasource should use the passed registry url. Datasources are using the default registry url, if the dependency doesn't define one.

So if the npm datasource extract the npmrc file, that code should be refactored, do that the npm manager extracts the npmrc file.

I think this should already be the case.

So the things to to here is to extract the registry url from the corresponding config file at manager extract call.

**Comment by rdicroce** [comment]
Like I said earlier in this thread, the problem with that design is that it doesn't address the case where the manager and the datasource are different. It works fine if the _npm_ manager generates a dependency, and then the npm datasource fetches that dependency. But it doesn't work if the _regex_ manager generates a dependency, and then the npm datasource fetches that dependency.

Unless you're suggesting that the regex manager should parse .yarnrc.yml/.npmrc/etc just like the npm datasource does? Seems messy to me since it results in duplicate logic. Even if you factor the logic into a single function that both datasources call, you still have to remember to make regex (and any similar datasources, like the one proposed in #6301) call the function.

**Comment by viceice** [comment]
This would be the tradeof we will do.

For example we use npm datasource with regex manager on some of our docker images and we don't want that the npm datasource will inherit npmrc settings.

So your tradeof is to configure registry urls manually for regex manager.

The npm manager can read and use the config to detect the correct registry urls.

**Comment by rarkins** [comment]
I agree: if using the regex manager and custom registries then it's a reasonable trade-off to manually configure packageRules to specify registryUrls

**Comment by rdicroce** [comment]
In that case, implementing this doesn't really save me any effort elsewhere. So I don't plan to work on it, and someone else can feel free to take this on if they want.

**Comment by rarkins** [comment]
I think let's close this issue as there's too much going on for anyone else to pick up.
